repository/src/fetch.rs

```diff
@@ -0,0 +1,269 @@
+                    .filter_map(|(commit, hash)| matches!(commit, Commit::Block(_)).then(|| *hash)),
+            )
+            .collect::<Vec<_>>();
+
+        // Try to finalized the last block with the known finalization proofs.
+        let mut hit = false;
+        for proof in &finalization_proofs {
+            if csv.verify_last_header_finalization(&proof.proof).is_ok() {
+                last_finalization_proof = proof.clone();
+                last_finalized_commit_hash = *block_commit_hashes.last().unwrap();
+                hit = true;
+            }
+        }
+
+        // If failed, finalize the last-last block.
+        if !hit && block_commit_hashes.len() > 2 {
+            let last_block = csv.get_block_headers().last().unwrap().clone();
+            last_finalization_proof = LastFinalizationProof {
+                proof: last_block.prev_block_finalization_proof.clone(),
+                height: last_block.height - 1,
+            };
+            last_finalized_commit_hash = block_commit_hashes[block_commit_hashes.len() - 2];
+        }
+    }
+
+    // step 2: create `a` and `b` branches.
+    for (remote_name, branch_name, commit_hash, commit) in next_step_branches {
+        let branch_displayed = format!(" {}/{}(at {})", remote_name, branch_name, commit_hash);
+        match &commit {
+            Commit::Agenda(agenda) => {
+                if agenda.height == last_finalization_proof.height + 1 {
+                    let branch_name = format!(
+                        "a-{:?}",
+                        commit
+                            .to_hash256()
+                            .to_string()
+                            .truncate(COMMIT_TITLE_HASH_DIGITS)
+                    );
+                    this.raw.create_branch(branch_name, commit_hash).await?;
+                }
+            }
+            Commit::AgendaProof(agenda_proof) => {
+                if agenda_proof.height == last_finalization_proof.height + 1 {
+                    let branch_name = format!(
+                        "a-{:?}",
+                        commit
+                            .to_hash256()
+                            .to_string()
+                            .truncate(COMMIT_TITLE_HASH_DIGITS)
+                    );
+                    this.raw.create_branch(branch_name, commit_hash).await?;
+                }
+            }
+            Commit::Block(block_header) => {
+                if block_header.height == last_finalization_proof.height + 1 {
+                    let branch_name = format!(
+                        "b-{:?}",
+                        commit
+                            .to_hash256()
+                            .to_string()
+                            .truncate(COMMIT_TITLE_HASH_DIGITS)
+                    );
+                    this.raw.create_branch(branch_name, commit_hash).await?;
+                }
+            }
+            x => warn!("incomplete remote branch {}: got {:?}", branch_displayed, x),
```

Comment:
```suggestion
            c => warn!("incomplete remote branch {}: got {:?}", branch_displayed, c),
```
how about c for a commit